src/lsm/tree.zig

```diff
@@ -618,37 +614,34 @@ pub fn TreeType(comptime Table: type, comptime Storage: type, comptime tree_name
             );
         }
 
-        fn compact_io(tree: *Tree) void {
+        fn compact_drive(tree: *Tree) void {
             assert(tree.compaction_io_pending <= 2 + tree.compaction_table.len);
             assert(tree.compaction_callback != null);
 
-            // Try to tick the cpu portion of the immutable table compaction:
+            // Always start one fake io_pending that is resolved right after
+            // to handle the case where this compaction tick triggers no IO.
+            // (For example, ticking the immutable table, or level B is already done).
+            tree.compaction_io_pending += 1;
+            defer tree.compact_tick_done();
+
+            // Try to tick the immutable table compaction:
             const compaction_beat = tree.compaction_op % config.lsm_batch_multiple;
             const even_levels = compaction_beat < half_measure_beat_count;
             if (even_levels) {
                 assert(tree.compaction_table_immutable.status == .idle);
             } else {
-                if (tree.compaction_table_immutable.status == .compacting) {
-                    tree.compact_io_tick(&tree.compaction_table_immutable);
-                }
+                tree.compact_tick(&tree.compaction_table_immutable);
             }
 
-            // Try to tick the cpu portion of the level compactions:
+            // Try to tick the compaction for each level:
             var it = CompactionTableIterator{ .tree = tree };
             while (it.next()) |context| {
-                if (context.compaction.status == .compacting) {
-                    assert(context.compaction.level_b == context.level_b);
-                    tree.compact_io_tick(context.compaction);
-                }
+                tree.compact_tick(context.compaction);
             }
-
-            // Always start one io_pending that is resolved in compact_cpu()
-            // to handle the case of no level or immutable table being selected for compaction
-            tree.compaction_io_pending += 1;
-            assert(tree.compaction_io_pending <= 2 + tree.compaction_table.len);
         }
 
-        fn compact_io_tick(tree: *Tree, compaction: anytype) void {
+        fn compact_tick(tree: *Tree, compaction: anytype) void {
+            if (compaction.status != .processing) return;
```

Comment:
If the compaction's status isn't `processing`, can we assert that it is `done`?

Reply:
Here, it could be `idle` if it was never started, or `done` if has fully completed. This function ticks a compaction if it's still `processing`. That status check can be moved to the caller though.

Reply:
I think it is ok where it is, I just didn't realize that a compaction could be `idle` here.